### PR TITLE
Add override constraining e to p ratios

### LINKED ISSUE #130: Add `stylised-storage` override
Previous studies constrained the energy to power ratios of battery and hydrogen. This constraint should be added as an optional override.

In [previous studies](https://github.com/timtroendle/money-land/blob/13c7caa6da34262461d158fcd85b82c069c772c7/src/template/model.yaml#L41), it looked like the following:

```yaml
overrides:
    stylised-storage:
        # Transform battery and hydrogen to stylised short term
        # and long term storage solutions.
        techs:
            battery:
                constraints:
                    energy_cap_per_storage_cap_min: 0.25 # max 4h full-power discharge
                costs:
                    monetary:
                        om_prod: 0 # ASSUME 0 instead of 0.000378 €/kWh, as the tiny number causes numerical issues
            hydrogen:
                constraints:
                    energy_cap_per_storage_cap_max: 0.25 # min 4h full-power discharge
```

### PR DESCRIPTION
**Pull request by timtroendle**
Fixes #130.

## Checklist

Any checks which are not relevant to the PR can be pre-checked by the PR creator. All others should be checked by the reviewer. You can add extra checklist items here if required by the PR.

- [x] CHANGELOG updated
- [x] Minimal workflow tests pass
- [x] Tests added to cover contribution
- [x] Documentation updated
- [x] Configuration schema updated

### CODE REVIEW

File: templates/README.md [hunk 1]
```diff
@@ -49,6 +49,11 @@ By default, euro-calliope contains a single technology for rooftop PV. This tech
 
 When using the `directional-rooftop-pv` override, there are three instead of just one technologies for rooftop PV. The three technologies comprise (1) south-facing and flat rooftops, (2) east- and west-facing rooftops, and (3) north-facing rooftops. This leads to higher capacity factors of rooftop PV as long as the potential of rooftop PV is not fully exploited. However, this also increases the complexity of the model.
 
+> exclusive-e-to-p-ratios
```
Comment (brynpickering) [inline templates/README.md:52]: `energy-to-power` isn't a massive increase in length, but makes the whole thing much more readable.

File: templates/README.md [hunk 2]
```diff
@@ -49,6 +49,11 @@ By default, euro-calliope contains a single technology for rooftop PV. This tech
...
+
+Constrains the energy to power ratios of battery and hydrogen storage in a way that they do not overlap. Battery storage is constrained to a ratio of ≤4h while hydrogen is constrained to a ratio of ≥4h. The ratio is derived from typical values of commercial lithium-ion batteries available today (2021). Constraining hydrogen storage as well ensures  it does not directly compete with battery storage but is used for longer durations of fours hours and longer.
```
Comment (brynpickering) [inline templates/README.md:54]: maybe mention that `energy` here refers to stored energy and `power` to charge/discharge capacity. Final sentence: `Constraining hydrogen storage as well ensures it does not directly compete with battery storage, but is used instead for durations of fours hours and longer.`.

File: templates/storage-techs.yaml [hunk 3]
```diff
@@ -23,3 +23,13 @@ techs:
         constraints:
             energy_cap_max: inf
             storage_cap_max: inf
+overrides:
+    exclusive-e-to-p-ratios:
+        # Ensures exclusive energy to power ratios of battery (≤4h) and hydrogen (≥4h) storage.
+        techs:
+            battery:
```
Comment (brynpickering) [inline templates/storage-techs.yaml:30]: These could be one-liners: `battery.constraints.energy_cap_per_storage_cap_min: 0.25`

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 * **ADD** automatic download of EEZ (#99).
 * **ADD** Danish Energy Agency and Schroeder et al (2013) cost data as well as no hydro fixed costs as optional overrides (#18, #129).
+* **ADD** override to constrain energy to power ratios of battery and hydrogen storage (#130).
```
Comment (brynpickering) [inline CHANGELOG.md:7]: Now it's directly in the middle of the others! Also, capitalisation needed.

**Review by brynpickering — APPROVED** [review]
Simple enough; all makes sense. Do you see this being a problem in the model, that batteries operate on cycles > a few days and hydrogen < 1 week? I would have thought that the self-discharge of batteries as well as their lower ratio of energy capacity to storage capacity costs would make it unattractive, without the need for an explicit constraint. The existence of this override would lead me (and I guess others) to believe that the model rarely behaves this way...

**Comment by timtroendle** [comment]
It's true, it may not be needed. But that depends on cost parameters of storage and other options.

I've simply always used it in the past and I always copy pasted it. That's why I am including it here. It doesn't really do harm, does it?

**Comment by brynpickering** [comment]
No, certainly no harm, except that we begin to creep up in the number of possible overrides and thus the complexity of the contents across the various YAML files. Moving forward we should probably consider whether to bake some of these things in with config params (i.e. the `battery` and `hydrogen` tech templates have this added to them) or to better structure the YAML files (as I mention in #137) so that it's easier to inspect the list of overrides and differentiate them from the base Calliope config.

**Pull request merged**